robot_framework/process.py

```diff
@@ -1,8 +1,242 @@
+    """Find the change date for the address of each person in the Excel file.
+
+    Args:
+        wb: The Excel workbook object.
+    """
+    ws: Worksheet = wb.active
+    columns = [c.value for c in ws[1]]
+
+    # Create "Adresseændringsdato" column
+    aendringsdato_index = len(columns) + 1
+    ws.cell(row=1, column=aendringsdato_index, value="Adresseændringsdato")
+    ws.column_dimensions[openpyxl.utils.get_column_letter(aendringsdato_index)].width = 21
+
+    # Get cpr column
+    cpr_index = columns.index('CPR') + 1
+
+    # Get dates from database
+    conn = pyodbc.connect('DSN=Datavarehuset')
+    cursor = conn.cursor()
+
+    # Get change date and write result to Excel file
+    for row_index in range(2, ws.max_row+1):
+        cpr = ws.cell(row_index, cpr_index).value
+        date = look_up_address_change(cursor, cpr)
+        if date:
+            c = ws.cell(row_index, aendringsdato_index, date)
+            c.number_format = 'dd-mm-yyyy'
+
+
+def look_up_address_change(cursor: pyodbc.Cursor, cpr: str) -> datetime:
+    """Look up the last time the given person moved address.
+    This is done by look at the address history and finding
+    the last time the address actually changed.
+
+    Args:
+        cursor: The pyodbc cursor object to use.
+        cpr: The cpr of the person to look up.
+
+    Returns:
+        _The date of the last address change if any.
+    """
+    query = "SELECT DatoFra, Adressenoegle FROM DWH.MART.AdresseHistorik WHERE CPR = ? ORDER BY DatoFra DESC"
+    cursor.execute(query, cpr)
+
+    if cursor.rowcount == 0:
+        return None
+
+    rows = cursor.fetchall()
+    date_from = rows[0].DatoFra
+    address_key = rows[0].Adressenoegle
+
+    # Iterate through the rows to find the oldest connected occurrence of the address
+    for row in rows:
+        if row.Adressenoegle == address_key:
+            date_from = row.DatoFra
+        else:
+            break
+
+    return date_from
+
+
+def calculate_difference(wb: openpyxl.Workbook):
+    """Calculate the difference in days from udbetalingsdato to adresseændring.
+    If the difference is less than 89 days the row is deleted.
+
+    Args:
+        wb: The Excel workbook object.
+    """
+    ws: Worksheet = wb.active
+
+    # Create new 'Difference' column
+    columns = [c.value for c in ws[1]]
+    difference_index = len(columns)+1
+    ws.cell(row=1, column=difference_index, value="Difference")
+    ws.column_dimensions[openpyxl.utils.get_column_letter(difference_index)].width = 12
+
+    # Get other columns
+    udbetalingsdato_index = columns.index('Udbetalingsdato')
+    aendringsdato_index = columns.index('Adresseændringsdato')
+
+    # Iterate rows backwards to delete as we go
+    for row_index in range(ws.max_row, 1, -1):
+        row = ws[row_index]
+        diff = row[aendringsdato_index].value - row[udbetalingsdato_index].value.date()
+
+        # Insert difference unless diff < 89 then delete row
+        if diff.days < 89:
```

Comment:
Where does the value `89` come from?

Reply:
The process description. Would you rather it was a config constant?